numpy/f2py/crackfortran.py

```diff
@@ -1425,53 +1425,52 @@ def analyzeline(m, case, line):
                 continue
-            llen = len(l[1])
-            for idx, v in enumerate(rmbadname(
-                    [x.strip() for x in markoutercomma(l[0]).split('@,@')])
-                                    ):
-                if v[0] == '(':
-                    outmess(
-                        'analyzeline: implied-DO list "%s" is not supported. Skipping.\n' % v)
+            for idx, v in enumerate(rmbadname([x.strip() for x in markoutercomma(l[0]).split('@,@')])):
+                if v.startswith('('):
+                    outmess('analyzeline: implied-DO list "%s" is not supported. Skipping.\n' % v)
                     # XXX: subsequent init expressions may get wrong values.
                     # Ignoring since data statements are irrelevant for
                     # wrapping.
                     continue
-                fc = 0
-                vtype = vars[v].get('typespec')
-                vdim = getdimension(vars[v])
-
-                if (vtype == 'complex'):
-                    cmplxpat = r"\(.*?\)"
-                    matches = re.findall(cmplxpat, l[1])
-                else:
-                    matches = l[1].split(',')
-
-                if v not in vars:
-                    vars[v] = {}
-                if '=' in vars[v] and not vars[v]['='] == matches[idx]:
-                    outmess('analyzeline: changing init expression of "%s" ("%s") to "%s"\n' % (
-                        v, vars[v]['='], matches[idx]))
-
-                if vdim is not None:
-                    # Need to assign multiple values to one variable
-                    vars[v]['='] = "(/{}/)".format(", ".join(matches))
-                else:
-                    vars[v]['='] = matches[idx]
+                if '!' in l[1]:
+                    # Fixes gh-24746 pyf generation
+                    # XXX: This essentially ignores the value for generating the pyf which is fine:
+                    # integer dimension(3) :: mytab
+                    # common /mycom/ mytab
+                    # Since in any case it is initialized in the Fortran code
+                    outmess('Comment line in declaration "%s" is not supported. Skipping.\n' % l[1])
+                    new_val = 0
+                    continue
+                try:
+                    vtype = vars[v].get('typespec')
+                    vdim = getdimension(vars[v])
+                    matches = re.findall(r"\(.*?\)", l[1]) if vtype == 'complex' else l[1].split(',')
+                    new_val = "(/{}/)".format(", ".join(matches)) if vdim else matches[idx]
+                except: # gh-24746
+                    idy, jdx, fc = 0, 0, 0
+                    while idy < len(l[1]) and (fc or l[1][idy] != ','):
+                        if l[1][idy] == "'":
+                            fc = not fc
+                        idy += 1
+                    new_val = l[1][jdx:idy]
+                    jdx = idy + 1
+                vars.setdefault(v, {})
```

Comment:
Is it possible to use `vars.setdefault(v, {})` before line 1454 `vtype = vars[v].get('typespec')`? Maybe in this case try...except block will not be needed and the regex part would enough for all cases.

Reply:
Yup, I've gone ahead and simplified it as per the excellent suggestions :)